src/pyarray.jl

```diff
@@ -0,0 +1,329 @@
+
+f_contiguous(i::PyArray_Info{T,N}) where {T,N} = f_contiguous(T, i.sz, i.st)
+c_contiguous(i::PyArray_Info{T,N}) where {T,N} =
+    f_contiguous(T, reverse(i.sz), reverse(i.st))
+
+
+#########################################################################
+# PyArray: no-copy wrapper around NumPy ndarray
+#
+# Hopefully, in the future this can be a subclass of StridedArray (see
+# Julia issue #2345), which will allow it to be used with most Julia
+# functions, but that is not possible at the moment.  So, to use this
+# with Julia linalg functions etcetera a copy is still required.
+
+"""
+    PyArray(o::PyObject)
+
+This converts an `ndarray` object `o` to a PyArray.
+
+This implements a nocopy wrapper to a NumPy array (currently of only numeric types only).
+
+If you are using `pycall` and the function returns an `ndarray`, you can use `PyArray` as the return type to directly receive a `PyArray`.
+"""
+mutable struct PyArray{T,N} <: AbstractArray{T,N}
+    o::PyObject
+    info::PyArray_Info
+    dims::Dims
+    st::NTuple{N,Int}
+    f_contig::Bool
+    c_contig::Bool
+    data::Ptr{T}
+
+    function PyArray{T,N}(o::PyObject, info::PyArray_Info) where {T,N}
+        if !aligned(info)
+            throw(ArgumentError("only NPY_ARRAY_ALIGNED arrays are supported"))
+        elseif !info.native
+            throw(ArgumentError("only native byte-order arrays are supported"))
+        elseif eltype(info) != T
+            throw(ArgumentError("inconsistent type in PyArray constructor"))
+        elseif length(info.sz) != N || length(info.st) != N
+            throw(ArgumentError("inconsistent ndims in PyArray constructor"))
+        end
+        return new{T,N}(o, info, tuple(info.sz...), div.(info.st, sizeof(T)),
+                        f_contiguous(info), c_contiguous(info),
+                        convert(Ptr{T}, info.data))
+    end
+end
+
+function PyArray(o::PyObject)
+    info = PyArray_Info(o)
+    return PyArray{eltype(info), length(info.sz)}(o, info)
+end
+
+size(a::PyArray) = a.dims
+ndims(a::PyArray{T,N}) where {T,N} = N
+
+similar(a::PyArray, T, dims::Dims) = Array{T}(undef, dims)
+
+"""
+Update the data ptr of the `a` to point to the buffer exposed by `o` through
+the Python buffer interface
+"""
+function setdata!(a::PyArray{T,N}, o::PyObject) where {T,N}
+    pybufinfo = a.info.pybuf
+    PyBuffer!(pybufinfo, o, PyBUF_ND_CONTIGUOUS)
+    dataptr = pybufinfo.buf.buf
+    a.data = reinterpret(Ptr{T}, dataptr)
+    a
+end
+
+function copy(a::PyArray{T,N}) where {T,N}
+    if N > 1 && a.c_contig # equivalent to f_contig with reversed dims
+        B = unsafe_wrap(Array, a.data, ntuple((n -> a.dims[N - n + 1]), N))
+        return permutedims(B, (N:-1:1))
+    end
+    A = Array{T}(undef, a.dims)
+    if a.f_contig
+        ccall(:memcpy, Cvoid, (Ptr{T}, Ptr{T}, Int), A, a, sizeof(T)*length(a))
+        return A
+    else
+        return copyto!(A, a)
+    end
+end
+
+# TODO: need to do bounds-checking of these indices!
+
```

Comment:
These `unsafe_load` calls are also not rooting `a` properly… in Julia 0.7 or later, they should all be of the form `GC.@preserve a unsafe_load(...)`

Reply:
I don't fully follow here, wouldn't user code always have a reference to the `a` that they're calling getindex on? So the array would be GC rooted that way?

Reply:
@JobJob I'm not sure how Julia's GC works but if I do `np.zeros(100)[1]` there is no Julia variable referencing the array so I guess it could be GC'ed away?

Reply:
Ahh, yeah that seems plausible, added a TODO